File: 백준/Gold/1240. 노드사이의 거리/노드사이의 거리.cc
// Authored by : chjh2129
#include <bits/stdc++.h>

using namespace std;

using pi = pair<int, int>;

int n, m;
vector<pi> adj[1005];
int dist[1005][1005];

void dfs(int prv, int cur, int root){
  for(auto &[nxt, weight] : adj[cur]){
    if(nxt == prv) continue;
    dist[root][nxt] = dist[root][cur] + weight;
    dfs(cur, nxt, root);
  }
}

int main(){
  cin.tie(nullptr)->sync_with_stdio(false);

  // input
  cin >> n >> m;
  for(int i = 0; i < n - 1; i++){
    int u, v, w;
    cin >> u >> v >> w;
    adj[u].push_back({v, w});
    adj[v].push_back({u, w});
  }

  // solve
  for(int i = 1; i <= n; i++) dfs(-1, i, i);
  
  // output
  while(m--){
    int u, v;
    cin >> u >> v;
    cout << dist[u][v] << '\n';
  }
}

File: 백준/Gold/1240. 노드사이의 거리/노드사이의 거리.py
#Authored by : chjh2129
import sys

input = sys.stdin.readline

def dfs(prv:int, cur:int, G:list[list[tuple]], D:list[int]) -> None:
  for nxt, weight in G[cur]:
    if nxt != prv:
      D[nxt] = D[cur] + weight
      dfs(cur, nxt, G, D)

def main() -> None:
  N, M = map(int, input().split())

  # variable
  G = [ list() for _ in range(N+1) ]
  D = [ [0 for _ in range(N+1)] for _ in range(N+1) ] 
  
  for _ in range(N-1):
    u, v, w = map(int, input().split())
    G[u].append((v, w))
    G[v].append((u, w))
  
  # solve
  for i in range(1, N+1):
    dfs(-1, i, G, D[i])

  # output
  for _ in range(M):
    u, v = map(int, input().split())
    print(D[u][v])
  
sys.exit(main())